cpp/include/cugraph/graph_mask_functions.cuh

```diff
@@ -0,0 +1,281 @@
+{
+  /**
+   *   1. For each vertex for each block, load the start and end offsets from indptr
+   *   2. Compute start and end indices in the mask, along w/ the start and end masks
+   *   3. For each element in the mask array, apply start or end mask if needed,
+   *      compute popc (or popcll) and perform summed reduce of the result for each vertex
+   */
+
+  typedef cub::BlockReduce<vertex_t, tpb> BlockReduce;
+
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+
+  int vertex = blockIdx.x;
+
+  vertex_t start_offset = indptr[vertex];
+  vertex_t stop_offset  = indptr[vertex + 1];
+
+  mask_type start_mask_offset = start_offset >> log_bits<mask_type>();
+  mask_type stop_mask_offset  = stop_offset >> log_bits<mask_type>();
+
+  vertex_t degree = 0;
+  for (vertex_t i = threadIdx.x; i <= (stop_mask_offset - start_mask_offset); i += tpb) {
+    mask_type mask_elm = edge_mask[i + start_mask_offset];
+
+    // Apply start / stop masks to the first and last elements
+    if (i == 0) {
+      mask_elm = mask_elm & std::numeric_limits<mask_type>::max()
+                              << bit_mod<mask_type>(start_offset);
+    }
+    if (i == stop_mask_offset - start_mask_offset) {
+      mask_elm =
+        mask_elm & std::numeric_limits<mask_type>::max() >>
+                     (std::numeric_limits<mask_type>::digits - bit_mod<mask_type>(stop_offset));
+    }
+
+    degree += safe_popc<mask_type>(mask_elm);
+  }
+
+  degree = BlockReduce(temp_storage).Sum(degree);
+
+  if (threadIdx.x == 0) { degrees_output[vertex_ids[vertex] - major_range_first] = degree; }
+}
+
+}  // namespace detail
+
+template <typename vertex_t, typename edge_t, typename mask_type>
+void masked_degrees(raft::handle_t const& handle,
+                    edge_t* degrees_output,
+                    vertex_t size,
+                    graph_mask_view_t<vertex_t, edge_t, mask_type> const& mask,
+                    edge_t const* indptr)
+{
+  CUGRAPH_EXPECTS(mask.get_edge_mask().has_value(), "Edge mask is required to compute degrees.");
+  detail::masked_degree_kernel<vertex_t, edge_t, mask_type, 128>
+    <<<size, 128, 0, handle.get_stream()>>>(
+      degrees_output, mask.get_edge_mask().value().data(), indptr);
+}
+
+template <typename vertex_t, typename edge_t, typename mask_type>
+void masked_degrees(raft::handle_t const& handle,
+                    edge_t* degrees_output,
+                    vertex_t size,
+                    graph_mask_view_t<vertex_t, edge_t, mask_type> const& mask,
+                    int major_range_first,
+                    vertex_t const* vertex_ids,
+                    edge_t const* indptr)
+{
+  CUGRAPH_EXPECTS(mask.get_edge_mask().has_value(), "Edge mask is required to compute degrees.");
+  detail::masked_degree_kernel<vertex_t, edge_t, mask_type, 128>
+    <<<size, 128, 0, handle.get_stream()>>>(
+      degrees_output, mask.get_edge_mask().value().data(), major_range_first, vertex_ids, indptr);
+}
```

Comment:
I assume this code is unnecessary if we roll-back the updates in neighbor sampling.

Ideally, all the complexities related to masking should be handled inside the primitives, and should not be exposed in non-detail space public API.